lib/ports.dart

```diff
@@ -289,13 +342,13 @@ class SingleResponseChannel<R> {
   /// If `onTimeout` is not provided either,
   /// the future is completed with `timeoutValue`, which defaults to `null`.
   SingleResponseChannel(
-      {FutureOr<R> Function(Null value) callback,
-      Duration timeout,
+      {FutureOr<R> Function(dynamic value)? callback,
+      Duration? timeout,
       bool throwOnTimeout = false,
-      FutureOr<R> Function() onTimeout,
-      R timeoutValue})
+      FutureOr<R> Function()? onTimeout,
+      R? timeoutValue})
       : _receivePort = RawReceivePort(),
-        _completer = Completer<R>.sync(),
+        _completer = Completer<R?>.sync(),
         _callback = callback,
         _zone = Zone.current {
```

Comment:
I'd consider throwing here if `R` is non-nullable and `onTimeout` is null, `throwOnTimeout` is false and `timeoutValue` is `null`.

That is:
```dart
    if (!throwOnError &&
        onTimeout == null &&
        timeoutValue == null &&
        timeoutValue is! R) {
      throw ArgumentError.value(null, "timeoutValue",
          "Value is needed and result type does not allow null");
```

Then I'd make `result` getter below have type `Future<R>` instead of `Future<R?>`.

Howeve, that doens't work with `singleResponseChannelWithoutTimeout', so more rewriting might be needed.

This is OK for now, but I might want to take another pass on the APIs before we release it.

Reply:
It's a huge condition, I hope that can be documented to describe how it works. 🤣

Reply:
It's not that bad. The documentation above already deals with it most of the cases, so it only needs to be stated in the bottom "otherwise" case where it says that `timeoutValue` is actually used.